src/write/encoder_string_writer.rs

```diff
@@ -0,0 +1,108 @@
+use crate::Config;
+use std::io;
+use std::io::Write;
+use super::encoder::EncoderWriter;
+
+/// A `Write` implementation that base64-encodes data using the provided config and accumulates the
+/// resulting base64 in memory, which is then exposed as a String via `into_inner()`.
+///
+/// # Examples
+///
+/// ```
+/// use std::io::Write;
+///
+/// let mut enc = base64::write::EncoderStringWriter::new(base64::STANDARD);
+///
+/// enc.write_all(b"asdf").unwrap();
+///
+/// // get the resulting String
+/// let b64_string = enc.into_inner();
+///
+/// assert_eq!("YXNkZg==", &b64_string);
+/// ```
+///
+/// # Panics
+///
+/// Calling `write()` (or related methods) or `finish()` after `finish()` has completed without
+/// error is invalid and will panic.
+///
+/// # Performance
+///
+/// B64-encoded data is buffered in the heap since the point is to collect it in a String.
+pub struct EncoderStringWriter {
+    encoder: EncoderWriter<Vec<u8>>,
+}
+
+impl EncoderStringWriter {
+    /// Create a new EncoderStringWriter that will encode with the provided config.
+    pub fn new(config: Config) -> EncoderStringWriter {
+        EncoderStringWriter::from(String::new(), config)
+    }
+
+    /// Create a new EncoderStringWriter that will append to the provided string.
+    pub fn from(s: String, config: Config) -> EncoderStringWriter {
```

Comment:
Taking by value is very restrictive. It would be much more continent if it took a `&mut String`. For example if you had a function that would append to a `&mut String` you couldn't call this. However you can always convert a `String` to a `&mut String`.

Reply:
I don't disagree it would be more convenient, but how would it be structured since I wouldn't be able own the underlying `Vec<u8>`? Perhaps with a new abstraction like `Utf8OnlyWriter` that could convert the incoming base64 bytes into a `str` to be `push_str()`'d onto the String?

Reply:
It does require `unsafe` but you can just instantiate a `EcoderWriter<&mut Vec<u8>>` by calling `as_mut_vec()`.

Reply:
:( I _really_ want to stay unsafe free. I'll see what I can do.

Reply:
I don't think it is possible then unless you make this class independent of `EncodeWriter`.

Actually you could probably make this the base class and push `char`s onto the string. Then `as_vec()` is safe and free. I don't know the performance of pushing `char` but I if you do `s.push(byte as char)` the compiler needs to notice that `byte < 128`. Actually doing it directly into a `String` is probably "more correct" since you are building a string and know that it will be valid.

Reply:
see `StrWriter`